tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc

```diff
@@ -0,0 +1,207 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifdef GOOGLE_CUDA
+#include "tensorflow/stream_executor/cuda/cuda_activation.h"
+#include "third_party/gpus/cuda/include/cuda.h"
+#include "third_party/gpus/cuda/include/cuda_runtime_api.h"
+#endif  // GOOGLE_CUDA
+
+#include "tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.h"
+
+#include "tensorflow/core/common_runtime/device/device_id_utils.h"
+#include "tensorflow/core/common_runtime/gpu/gpu_id.h"
+#include "tensorflow/core/common_runtime/gpu/gpu_init.h"
+#include "tensorflow/core/framework/allocator.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/util/env_var.h"
+
+namespace tensorflow {
+
+GPUcudaMallocAsyncAllocator::GPUcudaMallocAsyncAllocator(
+    PlatformGpuId platform_gpu_id, size_t pool_size,
+    bool reserve_memory, bool compute_stats)
+    : cuda_stream_(nullptr),
+      name_(absl::StrCat("gpu_async_", platform_gpu_id.value())),
+      pool_(nullptr) {
+  stream_exec_ =
+      DeviceIdUtil::ExecutorForPlatformDeviceId(GPUMachineManager(),
+                                                platform_gpu_id).ValueOrDie();
+
+#if CUDA_VERSION < 11020
+  LOG(FATAL) << "TF_GPU_ALLOCATOR=cuda_malloc_async need CUDA 11.2 or higher to compile.";
+#else
+
+  // WAR an CUDA 11.2 driver bug for multiple-GPU. It currently
+  // request that the context on GPU 0 is initialized. Which isn't the
+  // case for TF+horovod.
+  if (platform_gpu_id.value() > 0) {
+    CUcontext pctx; // We loose track of it. But this is fine.
+    CUresult err = cuDevicePrimaryCtxRetain(&pctx, 0);
+    if (err != CUDA_SUCCESS){
+      LOG(FATAL) << "Failed to create the context on device 0.";
+    }
+  }
+
+  se::cuda::ScopedActivateExecutorContext scoped_activation{stream_exec_};
+  int cuda_malloc_async_supported;
+  cudaDeviceGetAttribute(&cuda_malloc_async_supported,
+                         cudaDevAttrMemoryPoolsSupported,
+                         platform_gpu_id.value());
+  if (!cuda_malloc_async_supported) {
+    LOG(FATAL) << "TF_GPU_ALLOCATOR=cuda_malloc_async isn't currently supported."
+               << " Possible causes: device not supported, driver too old, "
+               << " OS not supported, CUDA version too old.";
+  }
+
+  cudaError_t cerr = cudaStreamCreate(&cuda_stream_);
```

Comment:
Consider adding a static factory function that initializes streams etc and returns a stream_executor::port::StatusOr. That way you can guarantee the object is initialized with non-nullable values

Reply:
This file use a very limited set of stream_executor funtionality as it would add too much overhead. So I think it make the file more clear to use the CUDA syntax everywhere. This make it more locally stable and to me it looks more readable then mixing 2 differents API.

If the creation fail, I launch a FATAL error 2 lines bellow. So there is no chance of having non-initialized object staying around.

Reply:
I don't view this as CUDA vs stream executor syntax as much as general readability/usability improvement. With something like a factory and non-nullable const members, a reader can more easily reason about the possible states that this object takes on. As it is currently written, the reader has to trace through the code to understand the possible states.

I'm not going to block the PR on this change because this can be easily updated later on.